### ISSUE TITLE
changeLanguage cause params to undefined

### ISSUE THREAD

**Issue by vendramini** [isaachinman/next-i18next#485]
## Describe the bug
When I call `changeLanguage` all my params added from `getInitialProps` disappear.

## Occurs in next-i18next version
"next-i18next": "^1.1.2",

## Steps to reproduce
```javascript
import React from 'react';
import {i18n, withTranslation} from '../i18n';

const _ = (
    props
) =>
{
    console.log(props);
    const t = props.t;
    const onClick = async () =>
    {
        await i18n.changeLanguage(i18n.language === 'en-US' ? 'pt-BR' : 'en-US');
    };

    return (
        <div>
            <p>{t('message')}</p>
            <button onClick={onClick}>trocar</button>
        </div>
    );
};

_.getInitialProps = async () =>
{
    return { namespacesRequired: ['common'] };
};

export default withTranslation('common')(_);
```
i18n.js
```javascript
const NextI18Next = require('next-i18next').default;

const languages = ['pt-BR', 'en-US'];
const i18n = new NextI18Next({
    defaultLanguage: 'en-US',
    otherLanguages: ['pt-BR'],
    fallbackLng: 'en-US',
    initImmediate: false,
});
i18n.i18n.languages = languages;

module.exports = i18n;
```

## Expected behaviour
All params remains the same.

## Screenshots
https://gofile.io/?c=ISD7VA

## OS (please complete the following information)
 - Device: Linux Mint 19.1.
 - Browser: Google Chrome | 76.0.3809.132 (Official Build) (64-bit)

## Additional context
My `userAgent` param comes from express. This package: https://github.com/biggora/express-useragent

But I don't think it's related, because even a hard-coded param like `namespacesRequired` disappears.

**Comment by vendramini** [comment]
I'm changing the language by cookies now and it's working very well:
```javascript
Cookies.set('next-i18next', i18n.language === 'en-US' ? 'pt-BR' : 'en-US', {expires: 30});
window.location.reload();
```
But it's not related with my issue, I choose this way to continue the development without lose params while waiting a response.

Thanks.

**Comment by isaachinman** [comment]
@vendramini Your second comment there about "changing language by cookies" is effectively negating the benefit of using a SPA in the first place.

The relevant code is [here](https://github.com/isaachinman/next-i18next/blob/master/src/hocs/app-with-translation.tsx#L46). Indeed we are not passing the query string along. I had considered whether or not to do this, and decided not to, because NextJs does not pass the query string along by default when you call `Router.replace`.

Happy to discuss the appropriateness of this functionality.

**Comment by vendramini** [comment]
Yes, I agree with you. Furthermore, changing language by cookies and reload means consume more from server. But at the time, it's the way I have to implement, because if I lose parameters my interface breaks - `userAgent` needs to exists to decide between iOS or Android UI.

I think a boolean value `changeLanguage(lang: string, withParams: boolean)` could solve the problem. We lose callback param, but we can return a Promise from changeLanguage and resolve it.

What do you think?

**Comment by isaachinman** [comment]
> What do you think?

I'd rather not start extending the i18n API. The scope of `next-i18next` is rather large as it is. I wonder if there are any creative solutions that don't involve adding new features?

**Comment by vendramini** [comment]
Understood.

Maybe a hook like `useParams` that copy the params first time, store internally then return it if  incoming params are undefined/null?

Something like:
```javascript
Component.getInitialProps = async props => 
{
    const params = useParams(props); //deep copy in client. For server, maybe a second argument defining which params to copy, to avoid copying req, res, etc.
    return {...params};
};
```
Does it makes sense for you? Just an idea.

**Comment by isaachinman** [comment]
Not sure I get your point. @StarpTech Any opinion here?

**Comment by StarpTech** [comment]
IMO it should be possible to change the language without loosing any state in order to produce the same page but with different language.

@isaachinman 
>  I had considered whether or not to do this, and decided not to, because NextJs does not pass the query string along by default when you call Router.replace.

This is correct as general rule because they can't know if the operations belong together in any way but we can. We previously relied on shallow routing where the state is [presevered](https://nextjs.org/docs#shallow-routing). We should do it here as well because we only [wanted](https://github.com/isaachinman/next-i18next/issues/305) to refetch getInitalProps.

**Comment by StarpTech** [comment]
I'm still not 100% sure if I understood the problem correctly.

> I had considered whether or not to do this, and decided not to, because NextJs does not pass the query string along by default when you call Router.replace.

We do it already by passing the full href to replace? :confused:

**Comment by isaachinman** [comment]
> I'm still not 100% sure if I understood the problem correctly.

Same. @vendramini Can you please clarify?

**Comment by vendramini** [comment]
I'm not sure if you asked me to clarify the problem or my suggestion.

The problem: losing params when changing language.
My suggestion: as you don't want to modify the API, we should solve the problem from outside. So I though in some kind of "cache-hook" to store the values. When shallow routing, the hook will be call and return the params.

But I added in the wrong way in my example above, it should be placed inside component, not getInitialProps. But as I said, was just an idea that appear quickly in my mind.

For my problem now I solved with redux.

**Comment by StarpTech** [comment]
@vendramini could you please provide a minimal reproducible example? It is incomplete. Where do you lose your params? I don't lose any params because the full url is passed to the router `replace` method.

**Comment by vendramini** [comment]
It's not the params from URL, it's from `getInitialProps`.

You can see in my video (https://gofile.io/?c=ISD7VA) that even the param from this lib `namespacesRequired` disappears.
```javascript
_.getInitialProps = async () =>
{
    return { namespacesRequired: ['common'] };
};
```

**Comment by isaachinman** [comment]
@vendramini There is no use of shallow routing at all in this package, anymore.

> But I don't think it's related, because even a hard-coded param

Have you tried removing it, then? We're not going to be able to help you without a full repository to reproduce.

**Comment by isaachinman** [comment]
Could be resolved by #509, but hard to say without fully understanding the user's problem here.

**Comment by Slowl** [comment]
I'm facing this issue and can't find to way to work around that ... I don't know if I'm doing something wrong here, I'll try to add more precisions  about the issue : 

Here is where I pass my data from my Express.js backend into my prop "data" :
```
homeTalent.getInitialProps = async({ query }) => ({
  namespacesRequired: ['homeTalent', 'navbar', 'downloadApp', 'footer'],
  data: query,
})
```

- On the first page load, everything works fine, and the data looks like this : 
![Capture du 2020-05-15 13-12-25](https://user-images.githubusercontent.com/20215160/82044626-f7da2a00-96ad-11ea-9c63-e48623fb4708.png)

- When I change the language with **i18n.changeLanguage('en')**, my data get overwritten like this : 
![Capture du 2020-05-15 13-17-25](https://user-images.githubusercontent.com/20215160/82045033-b007d280-96ae-11ea-86e5-5b74f2eee07c.png)

I would appreciate any help on this issue ! Thanks
